source/rpt/compartmentalization.cc

```diff
@@ -0,0 +1,963 @@
+        {
+          center = cell->center();
+          cells_and_indices.insert({cell->index(), cell});
+          points.push_back(center);
+          center.clear();
+        }
+    }
+
+  // output the center of cell to feed the Pyvista and extract the electric
+  // field and velocity at the cell centers
+  // from COMSOL or CFD simulation
+  std::string   filename = "cell_centers";
+  std::ofstream myfile;
+  std::string   sep = " ";
+  myfile.open(filename);
+  for (auto &point : points)
+    {
+      for (int j = 0; j < 3; j++)
+        {
+          myfile << point[j] << sep;
+        }
+      myfile << "\n";
+    }
+  myfile.close();
+}
+
+template <int dim>
+auto
+Compartmentalization<dim>::electric_field_stored_in_vector()
+  -> std::vector<std::vector<double>>
+{
+  std::vector<double> electric_field;
+  std::ifstream       fin("input_values_emw");
+  double              element;
+  while (fin >> element)
+    {
+      electric_field.push_back(element);
+    }
+
+
+  for (const auto &[cell_index, cell] : cells_and_indices)
+    {
+      std::vector<double> fill_matrix;
+      fill_matrix.push_back(cell_index);
+      fill_matrix.push_back(electric_field[cell_index]);
+      primer_matrix_index_emw.push_back(fill_matrix);
+      fill_matrix.clear();
+    }
+  return primer_matrix_index_emw;
+}
+
+template <int dim>
+auto
+Compartmentalization<dim>::electric_field_stored_in_map()
+  -> std::map<typename Triangulation<dim>::active_cell_iterator, double>
+{
+  // Read the electric field and write it in a vector
+  std::vector<double> electric_field;
+  std::ifstream       fin("input_values_emw");
+  double              element;
+  while (fin >> element)
+    {
+      electric_field.push_back(element);
+    }
+
+  // Fill the container with the electromagnetic field
+  for (const auto &[cell_index, cell] : cells_and_indices)
+    {
+      primer_map_cell_emw.insert({cell, electric_field[cell_index]});
+    }
+  return primer_map_cell_emw;
+}
+
+template <int dim>
+auto
+Compartmentalization<dim>::read_velocity_vector()
+  -> std::map<typename Triangulation<dim>::active_cell_iterator,
+              std::vector<double>>
+{
+  std::ifstream       inputFile{"velocity_vector"};
+  std::vector<double> velocity_components{
+    std::istream_iterator<double>{inputFile}, {}};
+  for (unsigned int i = 0; i < cells_and_indices.size(); i++)
+    {
+      std::vector<double> velocity_vector;
+      for (unsigned int j = 0; j < 3; j++)
+        {
+          velocity_vector.push_back(velocity_components[0]);
+          velocity_components.erase(velocity_components.begin());
+        }
+      // Push {cell,[vector of velocity]}
+      // in cells_and_indices[i], "i" is the key which is cell index
+      // cells_and_indices[i] in the world of C++ map, means the element which
+      // its key is "i"
+      map_of_all_cells_and_velocity_vectors.insert(
+        {cells_and_indices[i], velocity_vector});
+      velocity_vector.clear();
```

Comment:
I think instead of using map with the cell_id, you can use vectors with the index = the cell_id

Reply:
Also, you can use Tensor<1, dim> for the velocity components so the map (or vector if you choose to change it) will be a map/vector of Tensor